lib/Transforms/Instrumentation/GeneratePartialFnPass.cc

```diff
@@ -0,0 +1,162 @@
+#include "seahorn/Transforms/Instrumentation/GeneratePartialFnPass.hh"
+
+#include "seahorn/Analysis/SeaBuiltinsInfo.hh"
+#include "seahorn/InitializePasses.hh"
+#include "seahorn/Passes.hh"
+#include "seahorn/Support/SeaDebug.h"
+
+#include "llvm_seahorn/IR/LLVMContext.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/IRBuilder.h"
+
+#define PARTIAL_FN_ANNOTATION "partial"
+#define PARTIAL_FN_STUB_PREFIX "sea.synth.inferable."
+
+using namespace llvm;
+
+namespace {
+
+// Returns true if I is marked by the "partial" metadata.
+bool hasPartialAnnotation(const llvm::Instruction &I) {
+  if (auto *meta = I.getMetadata(llvm_seahorn::LLVMContext::MD_annotation)) {
+    auto *tuple = cast<MDTuple>(meta);
+    for (auto &N : tuple->operands()) {
+      if (cast<MDString>(N.get())->getString() == PARTIAL_FN_ANNOTATION) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+} // namespace
+
+namespace seahorn {
+
+// Delcared in GeneratePartialFnPass.h
+bool isInferable(const Function &F) {
+  const std::string prefix(PARTIAL_FN_STUB_PREFIX);
+  return (F.getName().str().compare(0, prefix.size(), prefix) == 0);
+}
+
+// Delcared in GeneratePartialFnPass.h
+bool isPartialFn(const Function &F) {
+  for (auto &BB : F) {
+    for (auto &I : BB) {
+      return hasPartialAnnotation(I);
```

Comment:
shouldn't this be more conservative. If a function is inlined, its instructions will all have annotations, and the annotations are likely preserved during inlining.

We could check that, for example, the entry basic block has annotations, or maybe that all instructions in the entry basic block have annotations.